lldb/source/Plugins/LanguageRuntime/Swift/SwiftMetadataCache.cpp

```diff
@@ -0,0 +1,306 @@
+#include "SwiftMetadataCache.h"
+
+#include "lldb/Core/DataFileCache.h"
+#include "lldb/Utility/DataEncoder.h"
+#include "lldb/Utility/LLDBLog.h"
+#include "lldb/Utility/Log.h"
+#include "llvm/CodeGen/AccelTable.h"
+#include "llvm/CodeGen/AsmPrinter.h"
+#include "llvm/Support/CachePruning.h"
+#include "llvm/Support/Compression.h"
+#include "llvm/Support/BLAKE3.h"
+
+using namespace lldb;
+using namespace lldb_private;
+using namespace swift::reflection;
+using namespace swift::remote;
+
+static DataFileCache *GetSwiftMetadataCache() {
+  if (!ModuleList::GetGlobalModuleListProperties().GetEnableSwiftMetadataCache())
+    return nullptr;
+
+  llvm::CachePruningPolicy policy;
+  ModuleListProperties &properties =
+      ModuleList::GetGlobalModuleListProperties();
+  policy.Interval = std::chrono::hours(1);
+  policy.MaxSizeBytes = properties.GetSwiftMetadataCacheMaxByteSize();
+  policy.Expiration =
+      std::chrono::hours(properties.GetSwiftMetadataCacheExpirationDays() * 24);
+  static DataFileCache *g_data_file_cache =
+      new DataFileCache(ModuleList::GetGlobalModuleListProperties()
+                            .GetSwiftMetadataCachePath()
+                            .GetPath(), policy);
+  return g_data_file_cache;
+}
+
+SwiftMetadataCache &SwiftMetadataCache::get() {
+  static SwiftMetadataCache instance;
+  return instance;
+}
+
+void SwiftMetadataCache::registerModuleWithReflectionInfoID(ModuleSP module,
+                                                            uint64_t info_id) {
+  Log *log = GetLog(LLDBLog::Types);
+  auto it = m_info_to_module.find(info_id);
+
+  // Since this is a static singleton, we might've registered and loaded the
```

Comment:
It means that since we may outlive `SwiftLanguageRuntime`, we may be asked to register the same module more than once (`SwiftLanguageRuntime` will call this function only once per module during its lifetime).

Reply:
I'll remove the singleton part of the comment though as that makes it more confusing.